src/mbgl/style/expression/index_of.cpp

```diff
@@ -0,0 +1,150 @@
+#include "mbgl/style/expression/expression.hpp"
+#include <mbgl/style/conversion_impl.hpp>
+#include <mbgl/style/expression/index_of.hpp>
+#include <mbgl/util/string.hpp>
+#include <iostream>
+namespace mbgl {
+namespace style {
+namespace expression {
+
+IndexOf::IndexOf(std::unique_ptr<Expression> keyword_, std::unique_ptr<Expression> input_, std::unique_ptr<Expression> fromIndex_) :
+        Expression(Kind::IndexOf, type::Number),
+        keyword(std::move(keyword_)),
+        input(std::move(input_)),
+        fromIndex(std::move(fromIndex_))
+    {}
+    
+
+EvaluationResult IndexOf::evaluate(const EvaluationContext &params) const {
+  const EvaluationResult evaluatedKeyword = keyword->evaluate(params);
+  const EvaluationResult evaluatedInput = input->evaluate(params);
+  if (!evaluatedKeyword) {
+    return evaluatedKeyword.error();
+  }
+  if (!evaluatedInput) {
+    return evaluatedInput.error();
+  }
+
+  if (!(evaluatedKeyword->is<double>() || evaluatedKeyword->is<std::string>() ||
+        evaluatedKeyword->is<bool>() || evaluatedKeyword->is<NullValue>())) {
+    return EvaluationError{"Expected first argument to be of type boolean, string, number or null, but found " + toString(typeOf(*evaluatedKeyword))  +   " instead."};
+  }
+
+  size_t fromIndexValue = 0;
+  if (fromIndex) {
+      const EvaluationResult evaluatedFromIndex = fromIndex->evaluate(params);
+    if (!evaluatedFromIndex) {
+        return evaluatedFromIndex.error();
+    }
+    if (!evaluatedFromIndex->is<double>()) {
+        return EvaluationError{"Expected third argument to be of type number, but found " + toString(typeOf(*evaluatedFromIndex))  + " instead."};
+    }
+    fromIndexValue = static_cast<size_t>(evaluatedFromIndex->get<double>());
```

Comment:
It was not clear (neither from JS) tests nor from documentation - I haven't looked into JS code though) what to do if `fromIndex` is out of array bounds: show error or just return -1. I decided to use latter option: it has easier implementation and feels more natural to me(e.g. in C++ `std::string::find` doesn't throw exception if `pos` parameter is out of bounds).